library/core/src/iter/sources/from_generator.rs

```diff
@@ -0,0 +1,28 @@
+use crate::ops::{Generator, GeneratorState};
+use crate::pin::Pin;
+
+/// Creates a new iterator where each iteration calls the provided generator.
+///
+/// Similar to [`iter::from_fn`].
+///
+/// [`iter::from_fn`]: crate::iter::from_fn
+#[inline]
+#[unstable(feature = "iter_from_generator", issue = "none", reason = "generators are unstable")]
+pub fn from_generator<G: Generator<Return = ()> + Unpin>(
```

Comment:
It might be reasonable to also add a version that accepts a `Pin<&mut G>`, since that way immovable generators can be used as well.

Reply:
I'm intrigued if there's a way to support both from the same function (is a `Pin<&mut Generator<Return = ()>>: Generator<Return = ()> + Unpin`?)

Either way, I think we can add that on a follow up PR.

Reply:
I'm not quite sure what you mean. Right now there isn't an easy way to accept both direct mutable references and pinned ones, but that would be a lovely thing.

Reply:
I'm just wondering what changes we could make so that `Pin<&mut Generator<Return = ()>>` can satisfy `Generator<Return = ()> + Unpin`, making this function work for both versions.

Reply:
Oh! Now I understand, I think. That's actually a good question...

Reply:
Pretty sure that's already the case for `Future` ([yup](https://doc.rust-lang.org/nightly/core/future/trait.Future.html#impl-Future-2)) so it shouldn't be hard to do it for `Generator`.

Reply:
Wait, no, [it has that already](https://doc.rust-lang.org/nightly/core/ops/trait.Generator.html#impl-Generator%3CR%3E-1)